### PR TITLE
Allow graph generators and conversion functions to be dispatched

### PR DESCRIPTION
**Pull request by eriknw**
For now, the `backend` argument *must* be used to dispatch to a backend. Global config for automatic conversions to backends *does not* work on graph generators, readers, converters, or any function that does not have any graph inputs. For example:
```python
G = nx.from_pandas_edgelist(df, backend="cugraph")
```
However, the backend function _will_ be used when testing with a backend. The backend graph will be converted to a networkx graph automatically in `_convert_and_call_for_tests`.

This follows up on #6840.

CC @aaronzo @MridulS @jim22k @rlratzel

### CODE REVIEW

**Comment by rlratzel** [comment]
I've made progress on some mockup scripts that demonstrate various dispatching use cases that I think can be supported incrementally over time.  They cover config, auto-conversions, fallback scenarios, multiple backends, dynamic graph updates, etc. and I think they could influence this PR. I'm hoping to make the mockups available this week and go over them at the next community call, or earlier if possible, with at least @eriknw and @MridulS .  Can we go over those before merging this? It might prevent some API churn if we decide we want to support things differently.

**Comment by rlratzel** [comment]
> Can we go over those before merging this? It might prevent some API churn if we decide we want to support things differently.

We had a good discussion on this topic recently.  We also brought it up at the last dispatching community call (some notes [here](https://hackmd.io/rqs_pWMxSLmICXCpI3w-Ug?view#August-31-2023)).  I think the outcome was that we need to continue the discussions and add an NXEP, but in the meantime, I don't think we should block this PR (assuming we can change things related to it in the future if necessary, which I think we can?).  I will add a link to the new NXEP for this once it's available.

cc @aaronzo @MridulS @jim22k

**Review by MridulS — COMMENTED** [review]
This looks good to me, but I couldn't really test this bit as graphblas doesn't implement from_pandas_edgelist or from_scipy_sparse_array

Not sure if I can setup nx-cugraph locally without GPUs

**Comment by eriknw** [comment]
Test added, which was a bit more squirrelly than I expected.

**Comment by eriknw** [comment]
Should I/we go ahead and decorate more graph creation functions?

**Comment by dschult** [comment]
The test seems to work. And it seems like a reasonable approach. So it looks good to go ahead and decorate more graph creation functions

File: networkx/algorithms/bipartite/generators.py
Comment (eriknw) [inline networkx/algorithms/bipartite/generators.py:442]: Should we use the dispatch name of `"bipartite_random_graph"` here? Should we be careful with any other names in `bipartite` submodule?

File: networkx/algorithms/tree/operations.py
Comment (eriknw) [inline networkx/algorithms/tree/operations.py:36]: FYI, I added this. The input types don't work well with our current dispatching machinery, but this allows `backend=whatever` to be used explicitly.

File: networkx/algorithms/bipartite/generators.py
Comment (dschult) [reply networkx/algorithms/bipartite/generators.py:442]: Yes -- names in the bipartite subpackage can clobber names in the main namespace. So we should handle bipartitie in a similar way to how It was handled for some algorithms like `approximations`.

**Review by dschult — APPROVED** [review]
This looks good to me...
Lot's here, but all pretty straightforward.

File: networkx/algorithms/bipartite/generators.py
Comment (eriknw) [reply networkx/algorithms/bipartite/generators.py:442]: Yeah, this PR currently changes the name of those that are clobbered, and so far our approach has been to _not_ preemptively add e.g. `"bipartite_"` or `"approximate_"` prefix. Following current precedent, we don't add `"bipartite_"` here even though `random_graph` seems kind of generic.

Comment (eriknw) [reply networkx/algorithms/bipartite/generators.py:442]: oh wait, you answered "yes"... should we add `"bipartite_"` prefix?

Comment (dschult) [reply networkx/algorithms/bipartite/generators.py:442]: You should treat it the same as you did for the algorithms. If you only did it for the names that actually clobber, then let's stick to that here too.
:)

**Review by MridulS — APPROVED** [review]

**Pull request merged**
